src/number_theory/fermat_psp.lean

```diff
@@ -0,0 +1,578 @@
+    have : m * k ∣ (j * k) ^ (n - 1) - 1 := by rwa [←hj, nat.div_mul_cancel hb],
+
+    -- Inequalities that will be useful later in the proof
+    have q₁ : 1 ≤ j * k,
+    { rw [←hj, nat.div_mul_cancel]; assumption },
+    have q₂ : 1 ≤ (j * k) ^ (n - 1) := one_le_pow_of_one_le q₁ (n - 1),
+    have q₃ : 1 ≤ n - 1,
+    { have : n - 1 + 1 = n := nat.sub_add_cancel (show 1 ≤ n, by linarith),
+      linarith },
+    have q₄ : (n - 1) - 1 + 1 = n - 1 := nat.sub_add_cancel q₃,
+
+    -- Follows from the fact that `m * k ∣ (j * k) ^ (n - 1) - 1`
+    have : k ∣ (j * k) ^ (n - 1) - 1 := dvd_of_mul_left_dvd this,
+
+    -- Follows from the fact that `2 ≤ n`
+    have q : k ∣ (j * k) ^ (n - 1),
+    { rw mul_pow,
+      apply dvd_mul_of_dvd_right,
+      rw [←q₄, pow_succ],
+      exact dvd.intro _ rfl },
+
+    -- Since we know that `k` divides both `(j * k) ^ (n - 1)` and `(j * k) ^ (n - 1) - 1`,
+    -- it must divide their difference, so `k ∣ 1`. This contradicts the assumption that `k` is a
+    -- prime.
+    have : k ∣ (j * k) ^ (n - 1) - ((j * k) ^ (n - 1) - 1) := nat.dvd_sub' q this,
+    rw nat.sub_sub_self q₂ at this,
+    exact nat.prime.not_dvd_one hk this },
+
+  -- If n = 1, then it follows trivially that n is coprime with b.
+  { have : n = 1 := by linarith,
+    rw this,
+    norm_num }
+end
+
+lemma probable_prime_iff_modeq (n : ℕ) {b : ℕ} (h : 1 ≤ b) :
+  probable_prime n b ↔ b^(n - 1) ≡ 1 [MOD n] :=
+begin
+  have : 1 ≤ b ^ (n - 1) := one_le_pow_of_one_le h (n - 1), -- For exact_mod_cast
+  split,
+  { intro h₁,
+    apply nat.modeq_of_dvd,
+    have h₂ : ↑n ∣ ↑(b ^ (n - 1)) - (1 : ℤ) := by exact_mod_cast h₁,
+    have h₃ : - (↑(b ^ (n - 1)) - (1 : ℤ)) = ((↑1 : ℤ) - ↑(b ^ (n - 1))) := by simp,
+    exact h₃ ▸ (dvd_neg _ _).mpr h₂ },
+  { intro h₁,
+    have h₂ : ↑n ∣ ↑(b ^ (n - 1)) - (1 : ℤ) := nat.modeq.dvd h₁.symm,
+    exact_mod_cast h₂ }
+end
+
+/--
+If `n` is a Fermat pseudoprime to base `b`, then `n` is coprime with `b`, assuming that `b` is
+positive.
+
+This lemma is a small wrapper based on `coprime_of_probable_prime`
+-/
+lemma coprime_of_fermat_psp (n b : ℕ) (h : fermat_psp n b) (h₁ : 1 ≤ b) : nat.coprime n b :=
+begin
+  cases h with hp hn,
+  cases hn with hn₁ hn₂,
+  exact coprime_of_probable_prime n b hp (by linarith) h₁,
+end
+
+/--
+All composite numbers are Fermat pseudoprimes to base 1.
+-/
+lemma pseudoprime_of_base_one (n : ℕ) (h₁ : 1 < n) (h₂ : ¬nat.prime n) : fermat_psp n 1 :=
+begin
+  split,
+  { have h : 0 = 1^(n - 1) - 1 := by norm_num,
+    show n ∣ 1^(n - 1) - 1, from h ▸ (dvd_zero n) },
+  { exact ⟨h₂, h₁⟩ }
+end
+
+-- For lemmas that are needed to prove statements in this file, but aren't directly related to
+-- Fermat pseudoprimes
+section helper_lemmas
+
+private lemma odd_of_prime_gt_two (p : ℕ) (h : nat.prime p) (hp : 2 < p) : odd p :=
+begin
+  rw [nat.odd_iff_not_even, even_iff_two_dvd],
+  rw nat.prime_dvd_prime_iff_eq nat.prime_two h,
+  exact ne_of_lt hp
```

Comment:
```suggestion
  exact hp.ne
```
This is a dot notation trick which is worth knowing.

Reply:
As mentioned above, `odd_of_prime_gt_two` has been replaced by `nat.prime.odd_of_ne_two`